controllers/capabilities/developeruser_controller.go

```diff
@@ -74,12 +81,61 @@ func (r *DeveloperUserReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 		reqLogger.V(1).Info(string(jsonData))
 	}
 
+	// DeveloperUser has been marked for deletion
+	if developerUserCR.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(developerUserCR, developerUserFinalizer) {
+		err = r.removeDeveloperUserFrom3scale(developerUserCR)
+		if err != nil {
+			r.EventRecorder().Eventf(developerUserCR, corev1.EventTypeWarning, "Failed to delete developer user", "%v", err)
+			return ctrl.Result{}, err
+		}
+
+		controllerutil.RemoveFinalizer(developerUserCR, developerUserFinalizer)
+		err = r.UpdateResource(developerUserCR)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
 	// Ignore deleted resource, this can happen when foregroundDeletion is enabled
 	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
 	if developerUserCR.DeletionTimestamp != nil {
 		return ctrl.Result{}, nil
 	}
 
+	if !controllerutil.ContainsFinalizer(developerUserCR, developerUserFinalizer) {
+		controllerutil.AddFinalizer(developerUserCR, developerUserFinalizer)
+		err = r.UpdateResource(developerUserCR)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	providerAccount, err := controllerhelper.LookupProviderAccount(r.Client(), developerUserCR.GetNamespace(), developerUserCR.Spec.ProviderAccountRef, r.Logger())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Retrieve ownersReference of tenant CR that owns the DeveloperUser CR
+	tenantCR, err := controllerhelper.RetrieveTenantCR(providerAccount, r.Client(), r.Logger(), developerUserCR.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If tenant CR is found, set it's ownersReference as ownerReference in the DeveloperUser CR
+	if tenantCR != nil {
+		updated, err := controllerhelper.SetOwnersReference(developerUserCR, r.Client(), tenantCR)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if updated {
+			return ctrl.Result{Requeue: true}, nil
+		}
+	}
+
```

Comment:
As `DeveloperUsers` are linked to `DeveloperAccounts`, should `DeveloperUsers` be cascade deleted to when the associated `DeveloperAccount` is deleted also ? :thinking:

Reply:
I think it makes all the sense. The `DeveloperUsers` have ownerrefs to `DeveloperAccounts`. And `DeveloperAccounts` have ownerrefs to `Tenant`.

The operator should make sure that when deleting a DevUser that if the devAccount does not exist in 3scale, it can be removed.